tsa-core/src/main/kotlin/org/usvm/machine/interpreter/TvmArtificialInstInterpreter.kt

```diff
@@ -280,52 +296,194 @@ class TvmArtificialInstInterpreter(
+                                sentMessages = messages,
+                                messageOrderNumber = 0,
+                            ),
+                        )
+                    }
+                }
 
-        val commitedState =
-            scope.calcOnState {
-                lastCommitedStateOfContracts[currentContract]
+                is ActionHandlingResult.RealFailure -> {
+                    this.calcOnState {
+                        val failure =
+                            TvmFailure(
+                                actionsHandlingResult.failure,
+                                TvmFailureType.UnknownError,
+                                phase,
+                                pathNode,
+                            )
+                        calcOnState {
+                            isExceptional = isExceptional || oldIsExceptional
+                        }
+                        newStmt(TsaArtificialExitInst(stmt.computePhaseResult, failure, lastStmt.location))
+                    }
+                }
+
+                is ActionHandlingResult.SoftFailure -> {
+                    this.calcOnState {
+                        val failure =
+                            TvmResult.TvmSoftFailure(
+                                actionsHandlingResult.failure,
+                                phase,
+                            )
+                        calcOnState {
+                            isExceptional = isExceptional || oldIsExceptional
+                        }
+                        newStmt(TsaArtificialExitInst(stmt.computePhaseResult, failure, lastStmt.location))
+                    }
+                }
             }
+        }
+    }
 
+    private fun visitParseActionInst(
+        scope: TvmStepScopeManager,
+        stmt: TsaArtificialActionParseInst,
+    ) {
         // temporarily remove isExceptional mark for further processing
         val oldIsExceptional = scope.calcOnState { isExceptional }
         scope.doWithState {
             isExceptional = false
         }
+        val (head, tail) =
+            stmt.yetUnparsedActions.splitHeadTail() ?: return run {
+                transactionInterpreter.resolveMessageReceivers(
+                    scope,
+                    stmt.parsedAndPreprocessedActions,
+                ) { parsingResult ->
+                    doWithState {
+                        newStmt(
+                            TsaArtificialHandleMessagesCostInst(stmt.computePhaseResult, stmt.location, parsingResult),
+                        )
+                    }
+                }
+            }
+        val tmpStmt = stmt.copy(yetUnparsedActions = tail)
+        val possibleParsedHeads =
+            transactionInterpreter.parseSingleActionSlice(scope, head, tmpStmt).getOrElse { return } ?: return
+        val someParsingSucceeded = ctx.mkOr(possibleParsedHeads.map { it.second })
+        scope.assert(someParsingSucceeded)
+            ?: return
```

Comment:
There shouldn't be the need to do this. `doWithConditions` should automatically kill current state, if no conditions could be met (if not, this is a bug). Now you add a redundant path constraint